### PR TITLE
Make NServiceBus handlers code context aware

### PR DESCRIPTION
**Pull request by timbussmann**
Summary:
* based on @danielmarbach Handler context PR #2669 this PR changes the interface for all message handler interfaces to receive an additional `IMessageHandlerContext` parameter.
* `IMessageHandlerContext` contains all operations from `IBus` which are related to the currently processed message (e.g. Reply, Headers, HandleLater, ...). These have been removed from `IBus`.
* Replacing `IBus` usages by `IMessageHandlerContext`
* Removed `IBus` dependency from `Saga`.
* Custom Sagafinders can no longer use reply or other message contextual operations

Connects to Particular/PlatformDevelopment#187

### CODE REVIEW

**Comment by timbussmann** [comment]
@Particular/nservicebus please review

**Comment by SzymonPobiega** [comment]
From the API point of view, it looks good. The only small change I would do is change the type of `MessageHeaders` in the `IMessageHandlerContext` to be a custom class which exposes two methods
 * `TryGet` that returns a header if present
 * `GetAll` which returns a collection of header objects, each containing the name and the value
The rationale would be to prevent handlers from using the headers collection to communicate with each other. The headers should be read-only when handling a message.

**Comment by danielmarbach** [comment]
I can review this as well this afternoon. Ok to wait until then?

On Monday, October 19, 2015, Szymon Pobiega <notifications@github.com>
wrote:

> From the API point of view, it looks good. The only small change I would
> do is change the type of MessageHeaders in the IMessageHandlerContext to
> be a custom class which exposes two methods
>
>    - TryGet that returns a header if present
>    - GetAll which returns a collection of header objects, each containing
>    the name and the value The rationale would be to prevent handlers from
>    using the headers collection to communicate with each other. The headers
>    should be read-only when handling a message.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Particular/NServiceBus/pull/3006#issuecomment-149123596>
> .
>


-- 
*#NSBcon 2015 • The "All About NServiceBus" conference is back *
*• Learn more <http://go.particular.net/l/27442/2015-08-21/5d4wzp> →*

**Comment by timbussmann** [comment]
@danielmarbach :+1: 
@SzymonPobiega I had a discussion with @andreasohlund about that header interface. I don't think a custom headers class would yield any benefit over exposing the headers as a IReadyOnlyDictionary which allows both options you mentioned and even LINQ capabilities for free. But on the discussion we decided to keep the interface as minimal as possible for now. I think both options are valuable.

**Comment by danielmarbach** [comment]
Agree with @timbussmann

On Monday, October 19, 2015, Tim Bussmann <notifications@github.com> wrote:

> @danielmarbach <https://github.com/danielmarbach> [image: :+1:]
> @SzymonPobiega <https://github.com/SzymonPobiega> I had a discussion with
> @andreasohlund <https://github.com/andreasohlund> about that header
> interface. I don't think a custom headers class would yield any benefit
> over exposing the headers as a IReadyOnlyDictionary which allows both
> options you mentioned and even LINQ capabilities for free. But on the
> discussion we decided to keep the interface as minimal as possible for now.
> I think both options are valuable.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Particular/NServiceBus/pull/3006#issuecomment-149148218>
> .
>


-- 
*#NSBcon 2015 • The "All About NServiceBus" conference is back *
*• Learn more <http://go.particular.net/l/27442/2015-08-21/5d4wzp> →*

File: src/NServiceBus.AcceptanceTests/Correlation/When_sending_with_no_correlation_id.cs [hunk 1]
```diff
@@ -35,15 +35,13 @@ public CorrelationEndpoint()
 
             public class MyResponseHandler : IHandleMessages<MyRequest>
             {
-                public Context Context { get; set; }
+                public Context TestContext { get; set; }
```
Comment (danielmarbach) [inline src/NServiceBus.AcceptanceTests/Correlation/When_sending_with_no_correlation_id.cs]: In the past we discussed that the `IMessageHandlerContext` becomes a sort of a locator for core provided services. From the perspective of the acceptance tests the `ScenarioContext` is a provided service. I think it would be good to start into that direction in another PR and make the `IMessageHandlerContext` extendable and the ATT framework can be the first user of this API. This makes the container bindings for `ScenarioContext` no longer necessary.

File: src/NServiceBus.Core.Tests/API/APIApprovals.ApproveNServiceBus.approved.txt [hunk 2]
```diff
@@ -501,18 +497,44 @@ namespace NServiceBus
...
+    public interface IMessageContext { }
     public interface IMessageCreator
     {
         T CreateInstance<T>();
         T CreateInstance<T>(System.Action<T> action);
         object CreateInstance(System.Type messageType);
     }
+    public interface IMessageHandlerContext
+    {
+        System.Collections.Generic.IDictionary<string, string> MessageHeaders { get; }
```
Comment (danielmarbach) [inline src/NServiceBus.Core.Tests/API/APIApprovals.ApproveNServiceBus.approved.txt]: `IReadOnlyDictionary`?

File: src/NServiceBus.AcceptanceTests/Correlation/When_sending_with_no_correlation_id.cs (hunk 1 above)
Comment (andreasohlund) [inline src/NServiceBus.AcceptanceTests/Correlation/When_sending_with_no_correlation_id.cs]: I like it

File: src/NServiceBus.Core.Tests/API/APIApprovals.ApproveNServiceBus.approved.txt [hunk 3]
```diff
@@ -2793,14 +2816,6 @@ namespace NServiceBus.Unicast
     [System.ObsoleteAttribute("Not a public API. Please use `MessageHandlerRegistry` instead. Will be removed in" +
         " version 7.0.0.", true)]
     public interface IMessageHandlerRegistry { }
-    public class MessageContext : NServiceBus.IMessageContext
```
Comment (danielmarbach) [inline src/NServiceBus.Core.Tests/API/APIApprovals.ApproveNServiceBus.approved.txt]: Should this be obsoleted?

File: src/NServiceBus.Core/IMessageHandlerContext.cs [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+
+        /// <summary>
+        /// The address of the endpoint that sent the current message being handled.
+        /// </summary>
+        string ReplyToAddress { get; }
+
+        /// <summary>
+        /// Gets the list of key/value pairs found in the header of the message.
+        /// </summary>
+        IDictionary<string, string> MessageHeaders { get; }
```
Comment (danielmarbach) [inline src/NServiceBus.Core/IMessageHandlerContext.cs]: `IReadOnlyDictionary`?

File: src/NServiceBus.Core/IMessageHandlerContextExtensions.cs [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+        /// Publish the message to subscribers.
+        /// </summary>
+        /// <param name="context">The instance of <see cref="IMessageHandlerContext"/> to use for the action.</param>
+        /// <param name="message">The message to publish.</param>
+        public static Task PublishAsync(this IMessageHandlerContext context, object message)
+        {
+            return context.PublishAsync(message, new PublishOptions());
+        }
+
+
```
Comment (danielmarbach) [inline src/NServiceBus.Core/IMessageHandlerContextExtensions.cs]: Unnecessary whitespace

**Comment by danielmarbach** [comment]
Do you squash before merge?

**Comment by timbussmann** [comment]
@danielmarbach yes, gonna squash it after all review changes have been made.

File: src/NServiceBus.Core.Tests/API/APIApprovals.ApproveNServiceBus.approved.txt (hunk 3 above)
Comment (timbussmann) [inline src/NServiceBus.Core.Tests/API/APIApprovals.ApproveNServiceBus.approved.txt]: the only way to work with the context was by accessing the (now deprecated) MessageContext property on IBus which returns the IMessageContext interface. So I'm not sure why that class was even public in the first place. I think users can't really access that class so it should be safe to just make it private?

Comment (andreasohlund) [inline src/NServiceBus.Core.Tests/API/APIApprovals.ApproveNServiceBus.approved.txt]: >I think users can't really access that class so it should be safe to just make it private?


I'd say to obsolete it anyway, doesn't cost us much and it will be semver compliant

File: src/NServiceBus.Core/MessageHandlerContext.cs [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+    class MessageHandlerContext : IMessageHandlerContext
+    {
+        ContextualBus bus;
+
+        public MessageHandlerContext(ContextualBus bus)
+        {
+            this.bus = bus;
+        }
+
+        public string MessageId => bus.MessageBeingProcessed.MessageId;
```
Comment (timbussmann) [inline src/NServiceBus.Core/MessageHandlerContext.cs]: @danielmarbach @andreasohlund I completely removed MessageContext and using the underlying `IncomingMessage` on the `ContextualBus` directly. What do you think?

File: src/NServiceBus.Core/MessageHandlerContext.cs [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+
+        public string ReplyToAddress => bus.MessageBeingProcessed.GetReplyToAddress();
+
+        public IReadOnlyDictionary<string, string> MessageHeaders => bus.MessageBeingProcessed.Headers;
```
Comment (timbussmann) [inline src/NServiceBus.Core/MessageHandlerContext.cs]: while this works, should this internally use a Lazy<IReadOnlyDictionary> to avoid newing that collection up every time you do a get?

Comment (andreasohlund) [inline src/NServiceBus.Core/MessageHandlerContext.cs]: I'd say premature optimization for now and note that we need to do indepth profiling before we release v6

File: src/NServiceBus.Core/MessageHandlerContext.cs (hunk 6 above)
Comment (andreasohlund) [inline src/NServiceBus.Core/MessageHandlerContext.cs]: :+1:

File: src/NServiceBus.Core/MessageHandlerContext.cs (hunk 7 above)
Comment (SimonCropp) [inline src/NServiceBus.Core/MessageHandlerContext.cs]: @timbussmann  but this doesnt new up a collection each time? it is just returning the message headers

**Comment by SzymonPobiega** [comment]
:+1: for squash and good to go!

File: src/NServiceBus.Core/MessageHandlerContext.cs (hunk 7 above)
Comment (timbussmann) [inline src/NServiceBus.Core/MessageHandlerContext.cs]: @SimonCropp I thought converting from Dictionary to ReadOnlyDictionary will new up a ReadOnlyDictionary and initialize it with the given dictionary. But that was just an assumption.

**Comment by timbussmann** [comment]
@danielmarbach @andreasohlund @SzymonPobiega squashed and good to go.
